src/uct/cuda/gdr_copy/gdr_copy_md.c

```diff
@@ -259,36 +300,246 @@ static uct_md_ops_t md_ops = {
 };
 
 static ucs_status_t
-uct_gdr_copy_md_open(uct_component_t *component, const char *md_name,
-                     const uct_md_config_t *config, uct_md_h *md_p)
+uct_gdr_copy_mem_reg_locked(uct_md_h uct_md, void *address, size_t length,
+                            const uct_md_mem_reg_params_t *params,
+                            uct_mem_h *memh_p)
 {
-    const uct_gdr_copy_md_config_t *md_config =
-                    ucs_derived_of(config, uct_gdr_copy_md_config_t);
     ucs_status_t status;
-    uct_gdr_copy_md_t *md;
 
-    md = ucs_malloc(sizeof(uct_gdr_copy_md_t), "uct_gdr_copy_md_t");
-    if (NULL == md) {
-        ucs_error("failed to allocate memory for uct_gdr_copy_md_t");
-        return UCS_ERR_NO_MEMORY;
+    pthread_mutex_lock(&uct_gdr_copy_context.lock);
+    status = uct_gdr_copy_mem_reg(uct_md, address, length, params, memh_p);
+    pthread_mutex_unlock(&uct_gdr_copy_context.lock);
+
+    return status;
+}
+
+static ucs_status_t
+uct_gdr_copy_mem_dereg_locked(uct_md_h uct_md,
+                              const uct_md_mem_dereg_params_t *params)
+{
+    ucs_status_t status;
+
+    pthread_mutex_lock(&uct_gdr_copy_context.lock);
+    status = uct_gdr_copy_mem_dereg(uct_md, params);
+    pthread_mutex_unlock(&uct_gdr_copy_context.lock);
+
+    return status;
+}
+
+static uct_md_ops_t md_locked_ops = {
+    .close              = uct_gdr_copy_md_close,
+    .query              = uct_gdr_copy_md_query,
+    .mkey_pack          = uct_gdr_copy_mkey_pack,
+    .mem_reg            = uct_gdr_copy_mem_reg_locked,
+    .mem_dereg          = uct_gdr_copy_mem_dereg_locked,
```

Comment:
since mom_reg/dereg is not a fast path, there's a cache in UCP, IMO we could use single ops struct and always take the global lock